---
title: "Ocean acidification and boring sponge on *Crassostrea virginica* GE"
date: "*Last run on `r format(Sys.time(), '%d %B %Y')`*"
output: 
  html_document:
    theme: simplex
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: hide
---

---

<style>
  h2{color: #DC7633 !important}
  h1{color: #5499C7 !important}
  body{background-color: white !important}
</style>

<style>
.list-group-item.active, .list-group-item.active:hover, .list-group-item.active:focus {
    color: #212F3D;
    background-color: #EDBB99;
    font-weight: bold;}
a {
    color: #5499C7;}
.nav-tabs > li.active > a, .nav-tabs > li.active > a:hover, .nav-tabs > li.active > a:focus {
    color: #DC7633;}
body {
    font-family: "Open Sans","Helvetica Neue",Helvetica,Arial,sans-serif;
    font-size: 13px;
    line-height: 1.42857143;
    color: #212F3D;}
</style>

---

```{r setup, include = FALSE} 

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
options(knitr.kable.NA = '')

library("knitr")

date <- Sys.Date() # For saving with the current date
set.seed(7) # set seed


## Setting standard theme for ggplot for all plots:
theme_bove <- function(base_size = 14) {
  theme_bw(base_size = base_size) %+replace%
    theme(
      # remove the gridlines
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      # remove formatting on background
      strip.background = element_blank(),
      panel.background = element_rect(fill = "transparent"),
      plot.background = element_rect(fill = "transparent", colour = NA),
      # modify legend theme
      legend.position = "none",
      legend.background = element_rect(fill = "transparent", colour = "transparent"),
      legend.key = element_rect(fill = "transparent", colour = "transparent")
    )
}


## Source the custom functions file:
#source("Code/CustomFunctions.R")

```

```{r install packages, eval = FALSE, include = FALSE}

### If any packages are not installed or need to be updated, you can look for them below:
## get gridSVG from github directly
# library(devtools) 
# devtools::install_github("cran/gridSVG")
# devtools::install_github('sinhrks/ggfortify')
# devtools::install_github("ropensci/rnaturalearthhires")

## get Bioconductor packages
if (!requireNamespace("BiocManager"))
install.packages("BiocManager")
BiocManager::install("DESeq2")

## installing WGCNA:
# source("http://bioconductor.org/biocLite.R")
BiocManager::install(c("AnnotationDbi", "impute", "GO.db", "preprocessCore"))
BiocManager::install("WGCNA", dependencies=TRUE)
BiocManager::install("arrayQualityMetrics", dependencies=TRUE) # use this arrayQualityMetrics install if using later versions of R (3.6.3 works)
#repos="http://cran.us.r-project.org"

## R version 3.6 is funky with arrayQualityMetrics so need this work around:
install.packages("ape", lib = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
library(ape, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
install.packages("magick", lib = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library", dependencies = FALSE)
library(magick, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
BiocManager::install("arrayQualityMetrics", type = "source", checkBuilt = TRUE, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")



### Used packages that need to be installed to run code (and can be sourced easier from CRAN):
needed_packages <- c("tidyverse", "dplyr", "readr", "plotly", "vegan", "data.table", "ggpubr", "pdftools", "ggrepel", "adegenet") # Specify necessary packages

not_installed <- needed_packages[!(needed_packages %in% installed.packages()[ , "Package"])] # Extract not installed packages
if(length(not_installed)) install.packages(not_installed) # Install not installed packages

```

```{r load packages, include=FALSE}

library(tidyverse)
library(dplyr)
library(arrayQualityMetrics) # need special install above
library(ggplot2)
library(readr)
library(plotly)
library(DESeq2) # need special install above
library(vegan)
library(data.table)
library(ggpubr)
library(pdftools)
library(ggrepel)
library(adegenet)

```


## Expoloring data {.tabset}

```{r read in data}

# read in the counts file
counts <- read.table("Data/CVIR_featurecounts_22Jun22.txt", header = TRUE, row.names = 1)
counts <- counts[c(-1:-5, -37)] # removing columns 1-5 since we do not need them for this (and the Undetermined column)
col_names <- colnames(counts)

# Remove some of the extra stuff in the column names to match with expDesign
col_names <- gsub("X", "", col_names)
col_names <- gsub("\\_S.*", "", col_names)
colnames(counts) <- col_names

# read in the experimental design .csv
expDesign <- read.csv("Data/cvir_expDesign.csv")
expDesign <- expDesign[match(col_names, expDesign$Sample_ID),] # reorder samples to match count df
expDesign$infect <- factor(expDesign$infect)
expDesign$pCO2 <- factor(expDesign$pCO2)
expDesign$Sample_ID <- factor(expDesign$Sample_ID)
expDesign$treat <- factor(paste(expDesign$infect, expDesign$pCO2, sep = "_"))

```

<br/>

### Filtering counts

```{r unfiltered size factor plots, fig.width = 10, fig.height = 4}

nrow <- nrow(counts) # number of rows/counts (38828)
countMat <- DESeqDataSetFromMatrix(counts, expDesign, ~ 1) # makes a DESeqDataSet object with count data, experimental design, and no design formula
counts_SF <- estimateSizeFactors(countMat) #  estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)

# make dataframe of the size factors to visualize
sizeFactors <- data.frame(sample = counts_SF@colData@listData[["Sample_ID"]], treat = counts_SF@colData@listData[["treat"]], sizeFactors = counts_SF@colData@listData[["sizeFactor"]])

# plot sizeFactors
ggplot(data = sizeFactors, aes(x = sample, y = sizeFactors, fill = treat)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = 2, type = "qual") +
  geom_bar(stat = "identity") +
  ggtitle("Unfiltered sizeFactors")

```

After viewing the size factors, it appears that we have 5 samples with really low values that should be removed:

* 19355
* 19372
* 19382
* 19496
* 19505

<br/>


```{r filter low reads (base mean 3)}

### Filtered counts and remove outliers (less than 3)

## remove rows with no or low counts (remove ones with base mean lower than 3)
keep <- rowSums(counts(counts_SF)) >= 3
counts_filter <- counts[keep,]
nrow_filter <- nrow(counts_filter)
# the updated number of rows (24384)


## outliers removed based on array quality metrics
outs <- c(14, 16, 18, 23, 24) # pull these outliers
filter_counts_out <- counts_filter[,-outs]
expDesign <- expDesign[-outs,]

```

```{r filtered size factor plots, fig.width = 10, fig.height = 4}}

## Redo the count matrix steps from above with the filtered data
filt_countMat <- DESeqDataSetFromMatrix(filter_counts_out, expDesign, ~ 1) # makes a DESeqDataSet object with count data, experimental design, and no design formula
counts_SF_filter <- estimateSizeFactors(filt_countMat) #  estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)


# make dataframe of the size factors to visualize
sizeFactors_filter <-  data.frame(sample = counts_SF_filter@colData@listData[["Sample_ID"]], treat = counts_SF_filter@colData@listData[["treat"]], sizeFactors = counts_SF_filter@colData@listData[["sizeFactor"]])

# plot sizeFactors
ggplot(data = sizeFactors_filter, aes(x = sample, y = sizeFactors, fill = treat)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = 2, type = "qual") +
  geom_bar(stat = "identity") +
  ggtitle("Filtered sizeFactors")

```

Without filtering low reads, we had a total of **`r nrow`** counts. After filtering out the low counts (those with a base mean less than 3), we now have **`r  nrow_filter`** counts remaining for all *C. virginica* samples.

<br/>
<br/>


### PCA plots of Global Gene Expression 

```{r dds model and save, eval = FALSE, echo=TRUE}

# going to use the filter_counts_out object created above because it is filtered for low reads and has outliers removed

## create the DESeq object
dds <- DESeqDataSetFromMatrix(countData = filter_counts_out, 
                              colData = expDesign,
                              design = ~ treat)
dds <- DESeq(dds) # differential expression analysis on gamma-poisson distribution
vsd <-varianceStabilizingTransformation(dds, blind = TRUE) # quickly estimate dispersion trend and apply a variance stabilizing transformation

## Save dds and vsd data 
save(dds, vsd, file = "Data/transformed_counts.RData")

```

```{r normalized count table}

## Load previously saved dds and vsd data 
load("Data/transformed_counts.Rdata")

## Normalize counts for table
dds <- estimateSizeFactors(dds)
norm_counts <- counts(dds, normalized = TRUE)

```

```{r GE PCA, include=FALSE}

### Updated 'plotPCA' (from DESeq2) to save all PCs for plasticity measures
plotPCA_allPCs <- function (object, intgroup = "condition", ntop = 500, returnData = FALSE) 
{
  rv <- rowVars(assay(object)) # Variance estimates for each row (column) in a matrix.
  select <- order(rv, decreasing = TRUE)[seq_len(min(ntop, length(rv)))] # orders genes from largest -> smallest and pulls ntop (default 500) genes
  pca <- prcomp(t(assay(object)[select, ]), center = TRUE, scale. = FALSE) # performs PCA on ntop (default 500) from dataset
  percentVar <- pca$sdev^2/sum(pca$sdev^2)
  if (!all(intgroup %in% names(colData(object)))) {
    stop("the argument 'intgroup' should specify columns of colData(dds)")
  }
  intgroup.df <- as.data.frame(colData(object)[, intgroup, drop = FALSE])
  group <- if (length(intgroup) > 1) {
    factor(apply(intgroup.df, 1, paste, collapse = ":"))
  }
  else {
    colData(object)[[intgroup]]
  }
  d <- data.frame(group = group, intgroup.df, name = colnames(object), pca$x)
  if (returnData) {
    attr(d, "percentVar") <- percentVar # add [1:2] if only want 1st two PC variances
    return(d)
  }
  ggplot(data = d, aes_string(x = "PC1", y = "PC2", color = "group")) + 
    geom_point(size = 3) + xlab(paste0("PC1: ", round(percentVar[1] * 
                                                        100), "% variance")) + ylab(paste0("PC2: ", round(percentVar[2] * 
                                                                                                            100), "% variance")) + coord_fixed()
}

## calculate PCs based on the vsd transformed GE
pcaData <- plotPCA_allPCs(vsd, intgroup = c("treat", "pCO2", "infect"), returnData = TRUE) # will provide all PCs
head(pcaData)

# extract variance for all PCs
percentVar <- round((100 * attr(pcaData, "percentVar")), 2)

```


```{r PERMANOVA}

# run the PERMANOVA with euclidean distances and 1500 iterations
head(pcaData[6:31])
GE_pca <- adonis2(pcaData[6:31] ~ pcaData$infect * pcaData$pCO2, method = 'eu', permutations = 1500)
GE_pca

```

<br/>


```{r GE PCA plot, fig.align='center', fig.width = 8, fig.height = 6}

## PCA: T
ggplot(pcaData, aes(x = PC1, y = PC2)) +
  geom_point(size = 3, aes(colour = treat),alpha = 0.8) +
  stat_ellipse(geom = "polygon", aes(colour = treat), fill = NA, type = "t", level = 0.95, show.legend = FALSE) +
  xlab(paste0("PC1 (", percentVar[1], "%)")) +
  ylab(paste0("PC2 (", percentVar[2], "%)")) +
  scale_colour_brewer(palette = 2, type = "qual") +
  theme_classic()

```

<br/>
<br/>


## Differential gene expression {.tabset}

<br/>

### 32°C

```{r 32C DESeq results}

res_32C_co2 <- results(dds, contrast = c("pCO2", "400", "2800"), alpha = 0.05)
res_32C_sponge <- results(dds, contrast = c("infect", "N", "S"), alpha = 0.05)
#res_32C_both <- results(dds_32C, contrast = c("timeF", "24", "3"), alpha = 0.05)

```

```{r 32C results summary, echo = TRUE}

summary(res_32C_co2)
summary(res_32C_sponge)
#summary(res_32C_both)

```

```{r 32C filter for DEGs}

## Filter for DEG per pairwise comparison (adjusted p value = 0.05)
# 32C at 0 v 3 hours
down32C_3hr <- row.names(res_32C_3hr[res_32C_3hr$padj < 0.05 & !is.na(res_32C_3hr$padj) & res_32C_3hr$log2FoldChange < 0, ])
up32C_3hr <- row.names(res_32C_3hr[res_32C_3hr$padj < 0.05 & !is.na(res_32C_3hr$padj) & res_32C_3hr$log2FoldChange > 0, ])

# 32C at 0 v 24 hours
down32C_24hr <- row.names(res_32C_24hr[res_32C_24hr$padj < 0.05 & !is.na(res_32C_24hr$padj) & res_32C_24hr$log2FoldChange < 0, ])
up32C_24hr <- row.names(res_32C_24hr[res_32C_24hr$padj < 0.05 & !is.na(res_32C_24hr$padj) & res_32C_24hr$log2FoldChange > 0, ])

# 32C at 3 v 24 hours
down32C_both <- row.names(res_32C_both[res_32C_both$padj < 0.05 & !is.na(res_32C_both$padj) & res_32C_both$log2FoldChange < 0, ])
up32C_both <- row.names(res_32C_both[res_32C_both$padj < 0.05 & !is.na(res_32C_both$padj) & res_32C_both$log2FoldChange > 0, ])



## Compile DEG dataframe
DEG_bar <- list("treat" = c("control vs 3hr", "control vs 24hr", "3hr vs 24hr"), 
                "up" = c(length(up32C_3hr), length(up32C_24hr), length(up32C_both)), 
                "down" = c((length(down32C_3hr) * -1), (length(down32C_24hr) * -1), (length(down32C_both) * -1)))

DEG_bar <- as.data.frame(DEG_bar)
DEG_bar <- gather(DEG_bar, reg, genes, up:down)
DEG_bar$treat <- ordered(DEG_bar$treat, levels = c("control vs 3hr", "control vs 24hr", "3hr vs 24hr"))

```

```{r 32C DEG plot}

## Plot bar graph of DEG

DEG_plot_32 <- ggplot(data = DEG_bar, aes(x = treat, y = genes, fill = reg)) +
  theme_classic() +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("#9ebcda", "#8c6bb1")) +
  geom_hline(yintercept = 0, linetype = "dashed", size=0.5) + 
  ylab("") +
  ylim(-500, 750) +  
  theme(legend.position = "none") +
  xlab ("") +
  ggtitle("32°C heat stress") 

```

<br/>
<br/>


### 37°C

```{r 37C DESeq results}

res_37C_3hr <- results(dds_37C, contrast = c("timeF", "3", "0"), alpha = 0.05)
res_37C_24hr <- results(dds_37C, contrast = c("timeF", "24", "0"), alpha = 0.05)
res_37C_both <- results(dds_37C, contrast = c("timeF", "24", "3"), alpha = 0.05)

```

```{r 37C results summary, echo = TRUE}

summary(res_37C_3hr)
summary(res_37C_24hr)
summary(res_37C_both)

```

```{r 37C filter for DEGs}

# 37C at 0 v 3 hours
down37C_3hr <- row.names(res_37C_3hr[res_37C_3hr$padj < 0.05 & !is.na(res_37C_3hr$padj) & res_37C_3hr$log2FoldChange < 0, ])
up37C_3hr <- row.names(res_37C_3hr[res_37C_3hr$padj < 0.05 & !is.na(res_37C_3hr$padj) & res_37C_3hr$log2FoldChange > 0, ])

# 37C at 0 v 24 hours
down37C_24hr <- row.names(res_37C_24hr[res_37C_24hr$padj < 0.05 & !is.na(res_37C_24hr$padj) & res_37C_24hr$log2FoldChange < 0, ])
up37C_24hr <- row.names(res_37C_24hr[res_37C_24hr$padj < 0.05 & !is.na(res_37C_24hr$padj) & res_37C_24hr$log2FoldChange > 0, ])

# 37C at 3 v 24 hours
down37C_both <- row.names(res_37C_both[res_37C_both$padj < 0.05 & !is.na(res_37C_both$padj) & res_37C_both$log2FoldChange < 0, ])
up37C_both <- row.names(res_37C_both[res_37C_both$padj < 0.05 & !is.na(res_37C_both$padj) & res_37C_both$log2FoldChange > 0, ])


## Compile DEG dataframe
DEG_bar_37 <- list("treat" = c("control vs 3hr", "control vs 24hr", "3hr vs 24hr"), 
                "up" = c(length(up37C_3hr), length(up37C_24hr), length(up37C_both)), 
                "down" = c((length(down37C_3hr) * -1), (length(down37C_24hr) * -1), (length(down37C_both) * -1)))

DEG_bar_37 <- as.data.frame(DEG_bar_37)
DEG_bar_37 <- gather(DEG_bar_37, reg, genes, up:down)
DEG_bar_37$treat <- ordered(DEG_bar_37$treat, levels = c("control vs 3hr", "control vs 24hr", "3hr vs 24hr"))

```

```{r 37C DEG plot}

## Plot bar graph of DEG

DEG_plot_37 <- ggplot(data = DEG_bar_37, aes(x = treat, y = genes, fill = reg)) +
  theme_classic() +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("#fec44f", "#ec7014")) +
  geom_hline(yintercept = 0, linetype = "dashed", size=0.5) + 
  ylim(-500, 750) +
  ylab("") +
  theme(legend.position = "none") +
  xlab ("") +
  ggtitle("37°C heat stress") 

```

<br/>

### DEGs by experiment

```{r DEG plot of both experiments, fig.align='center', fig.width = 9, fig.height = 4}

ggarrange(DEG_plot_32, DEG_plot_37)
ggsave("Figures/DEGs_perExperiment.png", width = 9, height = 4)
ggsave("Figures/DEGs_perExperiment.pdf", width = 9, height = 4)

```

```{r}

library(ggvenn)

## Make the lists of down v up regulated genes for each experiment
down_32 <- list("3 hrs" = down32C_3hr, "24 hrs" = down32C_24hr)
down_37 <- list("3 hrs" = down37C_3hr, "24 hrs" = down37C_24hr)

up_32 <- list("3 hrs" = up32C_3hr, "24 hrs" = up32C_24hr)
up_37 <- list("3 hrs" = up37C_3hr, "24 hrs" = up37C_24hr)


## Plot the seperate venn diagrams
ven_up_32 <- ggvenn(up_32, 
       show_percentage=FALSE,
       fill_color = c("#8073ac", "#2d004b"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#C8322F", set_name_size = 4,
       )

ven_up_37 <- ggvenn(up_37, 
       show_percentage=FALSE,
       fill_color = c("#e08214", "#7f3b08"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#C8322F", set_name_size = 4,
       )


ven_down_32 <- ggvenn(down_32, 
       show_percentage=FALSE,
       fill_color = c("#8073ac", "#2d004b"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#2B65AE", set_name_size = 4,
       )

ven_down_37 <- ggvenn(down_37, 
       show_percentage=FALSE,
       fill_color = c("#e08214", "#7f3b08"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#2B65AE", set_name_size = 4,
       )


## Combine and save plot
ggarrange(ven_up_32, ven_up_37, ven_down_32, ven_down_37)
ggsave("Figures/FigureXX_Vens.pdf", width = 4, height = 4)
ggsave("Figures/FigureXX_Vens.png", width = 4, height = 4)

```

  
## Session Information

Session information from the last full knit of Rmarkdown on `r format(Sys.time(), '%d %B %Y')`.

```{r session info}

sessionInfo()

```