---
title: "Ocean acidification and boring sponge on *Crassostrea virginica* GE"
date: "*Last run on `r format(Sys.time(), '%d %B %Y')`*"
output: 
  html_document:
    theme: simplex
    toc: yes
    toc_depth: 3
    toc_float: yes
    code_folding: hide
---

---

<style>
  h2{color: #DC7633 !important}
  h1{color: #5499C7 !important}
  body{background-color: white !important}
</style>

<style>
.list-group-item.active, .list-group-item.active:hover, .list-group-item.active:focus {
    color: #212F3D;
    background-color: #EDBB99;
    font-weight: bold;}
a {
    color: #5499C7;}
.nav-tabs > li.active > a, .nav-tabs > li.active > a:hover, .nav-tabs > li.active > a:focus {
    color: #DC7633;}
body {
    font-family: "Open Sans","Helvetica Neue",Helvetica,Arial,sans-serif;
    font-size: 13px;
    line-height: 1.42857143;
    color: #212F3D;}
</style>

---

```{r setup, include = FALSE} 

knitr::opts_chunk$set(echo = FALSE, warning = FALSE, message = FALSE)
options(knitr.kable.NA = '')

library("knitr")

date <- Sys.Date() # For saving with the current date
set.seed(7) # set seed


## Setting standard theme for ggplot for all plots:
theme_bove <- function(base_size = 14) {
  theme_bw(base_size = base_size) %+replace%
    theme(
      # remove the gridlines
      panel.grid.major = element_blank(),
      panel.grid.minor = element_blank(),
      # remove formatting on background
      strip.background = element_blank(),
      panel.background = element_rect(fill = "transparent"),
      plot.background = element_rect(fill = "transparent", colour = NA),
      # modify legend theme
      legend.position = "none",
      legend.background = element_rect(fill = "transparent", colour = "transparent"),
      legend.key = element_rect(fill = "transparent", colour = "transparent")
    )
}


## Source the custom functions file:
#source("Code/CustomFunctions.R")

```

```{r install packages, eval = FALSE, include = FALSE}

### If any packages are not installed or need to be updated, you can look for them below:
## get gridSVG from github directly
# library(devtools) 
# devtools::install_github("cran/gridSVG")
# devtools::install_github('sinhrks/ggfortify')
# devtools::install_github("ropensci/rnaturalearthhires")

## get Bioconductor packages
if (!requireNamespace("BiocManager"))
install.packages("BiocManager")
BiocManager::install("DESeq2")

## installing WGCNA:
# source("http://bioconductor.org/biocLite.R")
BiocManager::install(c("AnnotationDbi", "impute", "GO.db", "preprocessCore"))
BiocManager::install("WGCNA", dependencies=TRUE)
BiocManager::install("arrayQualityMetrics", dependencies=TRUE) # use this arrayQualityMetrics install if using later versions of R (3.6.3 works)
#repos="http://cran.us.r-project.org"

## R version 3.6 is funky with arrayQualityMetrics so need this work around:
install.packages("ape", lib = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
library(ape, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
install.packages("magick", lib = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library", dependencies = FALSE)
library(magick, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")
BiocManager::install("arrayQualityMetrics", type = "source", checkBuilt = TRUE, lib.loc = "/Library/Frameworks/R.framework/Versions/4.2/Resources/library")



### Used packages that need to be installed to run code (and can be sourced easier from CRAN):
needed_packages <- c("tidyverse", "dplyr", "readr", "plotly", "vegan", "data.table", "ggpubr", "pdftools", "ggrepel", "adegenet") # Specify necessary packages

not_installed <- needed_packages[!(needed_packages %in% installed.packages()[ , "Package"])] # Extract not installed packages
if(length(not_installed)) install.packages(not_installed) # Install not installed packages

```

```{r load packages, include=FALSE}

library(tidyverse)
library(dplyr)
library(arrayQualityMetrics) # need special install above
library(ggplot2)
library(readr)
library(plotly)
library(DESeq2) # need special install above
library(vegan)
library(data.table)
library(ggpubr)
library(pdftools)
library(ggrepel)
library(adegenet)
library(ggvenn)

source("Code/CustomFunctions.R")

```


## Expoloring data {.tabset}

```{r read in data}

# read in the counts file
counts <- read.table("Data/CVIR_featurecounts_22Jun22.txt", header = TRUE, row.names = 1)
counts <- counts[c(-1:-5, -37)] # removing columns 1-5 since we do not need them for this (and the Undetermined column)
col_names <- colnames(counts)

# Remove some of the extra stuff in the column names to match with expDesign
col_names <- gsub("X", "", col_names)
col_names <- gsub("\\_S.*", "", col_names)
colnames(counts) <- col_names

# read in the experimental design .csv
expDesign <- read.csv("Data/cvir_expDesign.csv")
expDesign <- expDesign[match(col_names, expDesign$Sample_ID),] # reorder samples to match count df
expDesign$infect <- factor(expDesign$infect)
expDesign$pCO2 <- factor(expDesign$pCO2)
expDesign$Sample_ID <- factor(expDesign$Sample_ID)
expDesign$treat <- factor(paste(expDesign$infect, expDesign$pCO2, sep = "_"))

```

<br/>

### Filtering counts

```{r unfiltered size factor plots, fig.width = 10, fig.height = 4}

nrow <- nrow(counts) # number of rows/counts (38828)
countMat <- DESeqDataSetFromMatrix(counts, expDesign, ~ 1) # makes a DESeqDataSet object with count data, experimental design, and no design formula
counts_SF <- estimateSizeFactors(countMat) #  estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)

# make dataframe of the size factors to visualize
sizeFactors <- data.frame(sample = counts_SF@colData@listData[["Sample_ID"]], treat = counts_SF@colData@listData[["treat"]], sizeFactors = counts_SF@colData@listData[["sizeFactor"]])

# plot sizeFactors
ggplot(data = sizeFactors, aes(x = sample, y = sizeFactors, fill = treat)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = 2, type = "qual") +
  geom_bar(stat = "identity") +
  ggtitle("Unfiltered sizeFactors")

```

After viewing the size factors, it appears that we have 5 samples with really low values that should be removed:

* 19355
* 19372
* 19382
* 19496
* 19505

<br/>


```{r filter low reads (base mean 3)}

### Filtered counts and remove outliers (less than 3)

## remove rows with no or low counts (remove ones with base mean lower than 3)
keep <- rowSums(counts(counts_SF)) >= 3
counts_filter <- counts[keep,]
nrow_filter <- nrow(counts_filter)
# the updated number of rows (24384)


## outliers removed based on array quality metrics
outs <- c(14, 16, 18, 23, 24) # pull these outliers
filter_counts_out <- counts_filter[,-outs]
expDesign <- expDesign[-outs,]

```

```{r filtered size factor plots, fig.width = 10, fig.height = 4}

## Redo the count matrix steps from above with the filtered data
filt_countMat <- DESeqDataSetFromMatrix(filter_counts_out, expDesign, ~ 1) # makes a DESeqDataSet object with count data, experimental design, and no design formula
counts_SF_filter <- estimateSizeFactors(filt_countMat) #  estimates the size factors using the "median ratio method" described by Equation 5 in Anders and Huber (2010)


# make dataframe of the size factors to visualize
sizeFactors_filter <-  data.frame(sample = counts_SF_filter@colData@listData[["Sample_ID"]], treat = counts_SF_filter@colData@listData[["treat"]], sizeFactors = counts_SF_filter@colData@listData[["sizeFactor"]])

# plot sizeFactors
ggplot(data = sizeFactors_filter, aes(x = sample, y = sizeFactors, fill = treat)) +
  theme_classic() +
  theme(axis.text.x = element_text(angle = 90)) +
  scale_fill_brewer(palette = 2, type = "qual") +
  geom_bar(stat = "identity") +
  ggtitle("Filtered sizeFactors")

```

Without filtering low reads, we had a total of **`r nrow`** counts. After filtering out the low counts (those with a base mean less than 3), we now have **`r  nrow_filter`** counts remaining for all *C. virginica* samples.

<br/>
<br/>


### PCA plots of Global Gene Expression 

```{r dds model and save, eval = FALSE, echo=TRUE}

# going to use the filter_counts_out object created above because it is filtered for low reads and has outliers removed

## create the DESeq object
dds <- DESeqDataSetFromMatrix(countData = filter_counts_out, 
                              colData = expDesign,
                              design = ~ treat)
dds <- DESeq(dds) # differential expression analysis on gamma-poisson distribution
vsd <-varianceStabilizingTransformation(dds, blind = TRUE) # quickly estimate dispersion trend and apply a variance stabilizing transformation

## Save dds and vsd data 
save(dds, vsd, file = "Data/transformed_counts.RData")

```

```{r normalized count table}

## Load previously saved dds and vsd data 
load("Data/transformed_counts.Rdata")

## Normalize counts for table
dds <- estimateSizeFactors(dds)
norm_counts <- counts(dds, normalized = TRUE)

```

```{r GE PCA, include=FALSE}

## calculate PCs based on the vsd transformed GE
pcaData <- plotPCA_allPCs(vsd, intgroup = c("treat", "pCO2", "infect"), returnData = TRUE) # will provide all PCs
#head(pcaData)

# extract variance for all PCs
percentVar <- round((100 * attr(pcaData, "percentVar")), 2)

```


```{r PERMANOVA}

# run the PERMANOVA with euclidean distances and 1500 iterations
#head(pcaData[6:31])
GE_pca <- adonis2(pcaData[6:31] ~ pcaData$infect * pcaData$pCO2, method = 'eu', permutations = 1500)
GE_pca

```

<br/>


```{r GE PCA plot, fig.align='center', fig.width = 8, fig.height = 6}

## PCA: treatment effects (crossed)
pca_plot <- ggplot(pcaData, aes(x = PC1, y = PC2)) +
  geom_point(size = 3, aes(colour = treat),alpha = 0.8) +
  stat_ellipse(geom = "polygon", aes(colour = treat), fill = NA, type = "t", level = 0.95, show.legend = FALSE) +
  xlab(paste0("PC1 (", percentVar[1], "%)")) +
  ylab(paste0("PC2 (", percentVar[2], "%)")) +
  scale_colour_brewer(palette = 2, type = "qual") +
  theme_classic()

pca_plot

ggsave("Figures/FigureXX_GEpca.pdf", width = 6, height = 4)
ggsave("Figures/FigureXX_GEpca.png", width = 6, height = 4)



```

PCA plot with each point being one oyster sample colored by treatment. Orange represents the control treatment (no boring sponge, pCO2 level 400), green represents no sponge, pCO2 level 2800, purple represents samples infected with boring sponge at pCO2 2800, and pink represents samples infected with boring sponge at pCO2 400.

<br/>
<br/>


## Differential gene expression {.tabset}

<br/>

### DEGs 

```{r general results of DEGs}

## Fist, we will look at the general 'results()' function
results(dds)
summary(results(dds))

```

```{r create results of DEGs by contrasts}

## When specifying your contrasts, make sure you 'control' level is the second value input and your 'treatment' is the first
res_con_sponge <- results(dds, contrast = c("treat", "S_400", "N_400"), alpha = 0.05)
res_con_pco2 <- results(dds, contrast = c("treat", "N_2800", "N_400"), alpha = 0.05)
res_trt <- results(dds, contrast = c("treat", "S_2800", "N_400"), alpha = 0.05)

#new
res_trt_pco2 <- results(dds, contrast = c("treat", "S_2800", "S_400"), alpha = 0.05)
res_stn_pco2 <- results(dds, contrast = c("treat", "S_400", "N_2800"), alpha = 0.05)
res_shn_pco2 <- results(dds, contrast = c("treat", "S_2800", "N_2800"), alpha = 0.05)

```

```{r results summary, echo = TRUE}

summary(res_con_sponge)
summary(res_con_pco2)
summary(res_trt)
summary(res_trt_pco2)
summary(res_stn_pco2)
summary(res_shn_pco2)

```

```{r filter for DEGs}

## Filter for DEG per pairwise comparison (adjusted p value = 0.05)
# sponge v no sponge at control pCO2
downCon_S <- row.names(res_con_sponge[res_con_sponge$padj < 0.05 & !is.na(res_con_sponge$padj) & res_con_sponge$log2FoldChange < 0, ])
upCon_S <- row.names(res_con_sponge[res_con_sponge$padj < 0.05 & !is.na(res_con_sponge$padj) & res_con_sponge$log2FoldChange > 0, ])

# pCO2 with no sponge
downTrt_N <- row.names(res_con_pco2[res_con_pco2$padj < 0.05 & !is.na(res_con_pco2$padj) & res_con_pco2$log2FoldChange < 0, ])
upTrt_N <- row.names(res_con_pco2[res_con_pco2$padj < 0.05 & !is.na(res_con_pco2$padj) & res_con_pco2$log2FoldChange > 0, ])

# control v double trt (sponge at high pCo2)
downTrt_S <- row.names(res_trt[res_trt$padj < 0.05 & !is.na(res_trt$padj) & res_trt$log2FoldChange < 0, ])
upTrt_S <- row.names(res_trt[res_trt$padj < 0.05 & !is.na(res_trt$padj) & res_trt$log2FoldChange > 0, ])

#sponge high vs sponge low
downtrtS <- row.names(res_trt_pco2[res_trt_pco2$padj < 0.05 & !is.na(res_trt_pco2$padj) & res_trt$log2FoldChange < 0, ])
uptrtS <- row.names(res_trt_pco2[res_trt_pco2$padj < 0.05 & !is.na(res_trt_pco2$padj) & res_trt$log2FoldChange > 0, ])

#sponge low vs no sponge high
downslsh <- row.names(res_stn_pco2[res_stn_pco2$padj < 0.05 & !is.na(res_stn_pco2$padj) & res_trt$log2FoldChange < 0, ])
upslsh <- row.names(res_stn_pco2[res_stn_pco2$padj < 0.05 & !is.na(res_stn_pco2$padj) & res_trt$log2FoldChange > 0, ])

#sponge high vs no sponge high
downsshn <- row.names(res_shn_pco2[res_shn_pco2$padj < 0.05 & !is.na(res_shn_pco2$padj) & res_trt$log2FoldChange < 0, ])
upsshn <- row.names(res_shn_pco2[res_shn_pco2$padj < 0.05 & !is.na(res_shn_pco2$padj) & res_trt$log2FoldChange > 0, ])

## Compile DEG dataframe
DEG_bar <- data.frame("treat" = c(toString(unlist(str_split(res_con_sponge@elementMetadata@listData[["description"]][2], " "))[c(8,6)]), toString(unlist(str_split(res_con_pco2@elementMetadata@listData[["description"]][2], " "))[c(8,6)]), toString(unlist(str_split(res_trt@elementMetadata@listData[["description"]][2], " "))[c(8,6)]), toString(unlist(str_split(res_trt_pco2@elementMetadata@listData[["description"]][2], " "))[c(8,6)]), toString(unlist(str_split(res_stn_pco2@elementMetadata@listData[["description"]][2], " "))[c(8,6)]), toString(unlist(str_split(res_shn_pco2@elementMetadata@listData[["description"]][2], " "))[c(8,6)])), 
                "up" = c(length(upCon_S), length(upTrt_N), length(upTrt_S), length(uptrtS), length(upslsh), length(upsshn)), 
                "down" = c((length(downCon_S) * -1), (length(downTrt_N) * -1), (length(downTrt_S) * -1), (length(downtrtS) * -1), (length(downslsh) * -1), (length(downsshn) * -1)))

DEG_bar <- gather(DEG_bar, reg, genes, up:down)

split_stringfunc <- function(x) {
  toString(unlist(str_split(x@elementMetadata@listData[["description"]][2], " "))[c(8,6)])
}

split_stringfunc(res_trt)

```

```{r DEG plot}

## Plot bar graph of DEG

ggplot(data = DEG_bar, aes(x = treat, y = genes, fill = reg)) +
  theme_classic() +
  geom_bar(stat="identity") +
  scale_fill_manual(values = c("cadetblue2", "coral")) +
  geom_hline(yintercept = 0, linetype = "dashed", size=0.5) + 
  ylab("Sig DEGs") +
  theme(legend.position = "none", axis.text.x = element_text(angle = 90)) +
  xlab ("")

ggsave("Figures/FigureXX_DEGsBar.pdf", width = 6, height = 4)
ggsave("Figures/FigureXX_DEGsBar.png", width = 6, height = 4)

```

<br/>
<br/>


### Paired Venns of DEGs

```{r paired Venns}

## Comparing up NS_400 vs S_2800 and up S_2800 vs NS_2800
venn_loci_df <- list("NS_400 vs S_2800" = upTrt_S, "S_2800 vs NS_2800" = upsshn)

## Comparing up NS_400 vs S_2800 and up S_2800 vs NS_2800
venn_loci_up <- list("NS_400 vs S_2800" = upTrt_S, "S_2800 vs NS_2800" = uptrtS)

## Comparing down NS_400 vs S_2800 and up S_2800 vs NS_2800
venn_loci_down <- list("NS_400 vs S_2800" = downTrt_S, "S_2800 vs NS_2800" = downsshn)

## Comparing down NS_400 vs S_2800 and down S_2800 vs S_400
venn_loci2_down <- list("NS_400 vs S_2800" = downTrt_S, "S_2800 vs NS_2800" = downtrtS)

## Plot the seperate venn diagrams
ven1 <- ggvenn(venn_loci_df, 
       show_percentage=FALSE,
       fill_color = c("darkorange", "coral"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#C8322F", set_name_size = 3,
       )

ven2 <- ggvenn(venn_loci_up, 
       show_percentage=FALSE,
       fill_color = c("darkorange", "coral"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#C8322F", set_name_size = 3,
       )

ven3 <- ggvenn(venn_loci_down, 
       show_percentage=FALSE,
       fill_color = c("cyan4", "cornflowerblue"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "blue", set_name_size = 3,
       )

ven4 <- ggvenn(venn_loci2_down, 
       show_percentage=FALSE,
       fill_color = c("cyan4", "cornflowerblue"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "blue", set_name_size = 3,
       )


## Combine and save plot
ggarrange(ven1, ven2, ven3, ven4)
ggsave("Figures/FigureXX_PairedVens.pdf", width = 4, height = 4)
ggsave("Figures/FigureXX_PairedVens.png", width = 4, height = 4)

```
<br/>
<br/>

### Full Venns of DEGs

Need to add all treatments with DEGs to the below code (ANGELA)

```{r full venns}

## Venn of up DEGs
venn_up <- list("NS_400 vs S_2800" = upTrt_S, "S_2800 vs NS_2800" = uptrtS, "S_2800 vs NS_2800" = upsshn, "N_2800 vs S_400" = upslsh)

## Comparing down NS_400 vs S_2800 and up S_2800 vs NS_2800
venn_down <- list("NS_400 vs S_2800" = downTrt_S, "S_2800 vs NS_2800" = downsshn, "S_2800 vs NS_2800" = downtrtS)


## Plot the seperate venn diagrams
ven_up <- ggvenn(venn_up, 
       show_percentage=FALSE,
       fill_color = c("darkorange", "coral", "red", "cornsilk2"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "#C8322F", set_name_size = 2,
       )

ven_down <- ggvenn(venn_down, 
       show_percentage=FALSE,
       fill_color = c("cyan4", "cornflowerblue", "darkslategray3"), fill_alpha = 0.3,
       stroke_size = 0.4, text_size = 3, text_color = "#25292B",
       set_name_color = "blue", set_name_size = 3,
       )

## Combine and save plot
ggarrange(ven_up, ven_down)
ggsave("Figures/FigureXX_fullVens.pdf", width = 3, height = 2)
ggsave("Figures/FigureXX_fullVens.png", width = 3, height = 2)

```

<br/>
<br/>


## Gene expression plasticity {.tabset}

```{r}

mycolors <- c("coral", "darkorange", "red")

#### here is the code from the function without modifications:
## To run the function, enter the following objects:
# PCAplast(pca = XXX, # the PCA dataframe containing the PCA eigenvalues
#          data = XXX, # the condition/treatment data corresponding to samples
#          sample_ID = "XXX", # the name of column that provide unique ID per sample (if blank, will pull rownames for this)
#          num_pca =  "XXX", # the number of PCAs to include in analysis (default is 'all', but you can specify another number with a minimum of 2 PCAs)
#          control_col = "XXX", # what the 'treatment' column is called
#          control_lvl = "XXX", # control level of the treatment. If blank, a control mean per control level is assumed
#          group = "XXX") # the grouping column (i.e., colony). If blank, will assume control level grouping only!




## To run the function, enter the following objects:
ge_plast_all <-  PCAplast(pca = pcaData[6:31], # the PCA dataframe containing the PCA eigenvalues
                      data = pcaData[-6:-31], # the condition/treatment data corresponding to samples
                      num_pca =  "all", # the number of PCAs to include in analysis (default is 'all', but you can specify another number with a minimum of 2 PCAs)
                      control_col = "treat", # what the 'treatment' column is called
                      control_lvl = "N_400") # control level of the treatment. If blank, a control mean per level is assumed

#plasticity plot 1
pca_all_plot <- ggplot(data = ge_plast_all, aes(x = treat, y = dist, colour = treat)) + 
  geom_point(position = position_dodge(width = 0.5)) + 
  geom_boxplot(alpha = 0.5, color = mycolors) +
  xlab("Treatment") +
  ggtitle("B)")


ge_plast_2 <-  PCAplast(pca = pcaData[6:31], # the PCA dataframe containing the PCA eigenvalues
                      data = pcaData[-6:-31], # the condition/treatment data corresponding to samples
                      num_pca =  "all", # the number of PCAs to include in analysis (default is 'all', but you can specify another number with a minimum of 2 PCAs)
                      control_col = "treat", # what the 'treatment' column is called
                      control_lvl = "N_400") # control level of the treatment. If blank, a control mean per level is assumed

#plasticity plot2
pca2_plot <- ggplot(data = ge_plast_2, aes(x = treat, y = dist, colour = treat)) + 
  geom_point(position = position_dodge(width = 0.5)) + 
  geom_boxplot(alpha = 0.5) +
  xlab("Treatment") +
  ggtitle("A)")

# ANOVA test
res_aov <- aov(dist ~ treat,
  data = ge_plast_2
)
summary(res_aov)

# ANOVA test ge_plast_all
res_aov_all <- aov(dist ~ treat,
  data = ge_plast_all
)
summary(res_aov_all)

# calculate mean and standard error for ge_plast_1
plasticity1_means_se <- ge_plast_all %>%  
  group_by(treat) %>% 
  summarize(mean_N=mean(dist), 
            sd_N=sd(dist), 
            N_N=n(), 
            se=sd_N/sqrt(N_N)
  ) 

# calculate mean and standard error for ge_plast_2
plasticity_means_se <- ge_plast_2 %>%  
  group_by(treat) %>% 
  summarize(mean_N=mean(dist), 
            sd_N=sd(dist), 
            N_N=n(), 
            se=sd_N/sqrt(N_N)
  ) 

#plot means and standard error
mean_se_plot <- ggplot(plasticity_means_se, aes(x=treat, y=mean_N, colour = treat)) + 
  geom_point() +
  geom_errorbar(aes(ymin=mean_N-se, ymax=mean_N+se), width=0.2, size=0.5) +
  ylab("Mean of Dist") +
  ggtitle("C)")

mean_se_plot_all <- ggplot(plasticity1_means_se, aes(x=treat, y=mean_N, colour = treat)) + 
  geom_point() +
  geom_errorbar(aes(ymin=mean_N-se, ymax=mean_N+se), width=0.2, size=0.5) +
  ylab("Mean of Dist") +
  ggtitle("D)")

ggarrange(pca2_plot, pca_all_plot, mean_se_plot, mean_se_plot_all)


```

Assessment of plasticity (calculated as the average of the distances of each sample to the mean eigenvalue) in different treatments. N_400 was omitted due to _____. 

## Results
There is no statistically significant difference in the means of plasticity across different treatments (p = 0.17). Overall, samples in high pCO2 (2800) and infected with sponge exhibited the highest plasticity. 

<br/>
<br/>
  
## Session Information

Session information from the last full knit of Rmarkdown on `r format(Sys.time(), '%d %B %Y')`.

```{r session info}

sessionInfo()

```